File: sdk/js/examples/cancellable-download.ts
// -------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
// -------------------------------------------------------------------------

/**
 * Example: Cancelling a model download using Foundry Local JS/TS SDK.
 *
 * Demonstrates how to cancel a running model download using the standard
 * `AbortController` / `AbortSignal` pattern.
 */

import { FoundryLocalManager } from '../src/index.js';

async function main() {
    try {
        // 1. Initialize the SDK
        console.log('Initializing Foundry Local SDK...');
        const manager = FoundryLocalManager.create({
            appName: 'CancellableDownloadExample',
            logLevel: 'info'
        });
        console.log('✓ SDK initialized');

        // Register execution providers
        const epResult = await manager.downloadAndRegisterEps();
        console.log(`EP registration success: ${epResult.success}`);

        // 2. Pick a model
        const modelAlias = 'phi-4-mini';
        const model = await manager.catalog.getModel(modelAlias);
        if (!model) {
            throw new Error(`Model ${modelAlias} not found in catalog`);
        }

        if (model.isCached) {
            console.log(`Model '${modelAlias}' is already cached. Remove it first to test cancellation.`);
            return;
        }

        // 3. Create an AbortController
        const controller = new AbortController();

        // 4. Start the download (returns a promise)
        console.log('Starting download...');
        const downloadPromise = model.download(
            (progress: number) => {
                process.stdout.write(`\rDownload: ${progress.toFixed(1)}%`);
            },
            controller.signal  // pass the AbortSignal
        );

        // 5. Cancel after 3 seconds
        console.log('Will cancel in 3 seconds...');
        setTimeout(() => {
            console.log('\nCancelling download...');
            controller.abort();
        }, 3000);

        // 6. Await the result
        await downloadPromise;
        console.log('\nDownload completed before cancellation took effect.');

    } catch (error) {
        if (error instanceof Error && error.message === 'Operation cancelled') {
            console.log(`\nDownload was cancelled: ${error.message}`);
        } else {
            console.error('Error:', error);
            process.exit(1);
        }
    }
}

main().catch(console.error);

export { main };

File: sdk/python/examples/cancellable_download.py
#!/usr/bin/env python3
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------

"""Example: Cancelling a model download using Foundry Local Python SDK.

Demonstrates how to start a model download on a background thread and
cancel it from the main thread using ``threading.Event``.
"""

import threading
import time

from foundry_local_sdk import Configuration, FoundryLocalManager


def main():
    # 1. Initialize the SDK
    config = Configuration(app_name="CancellableDownloadExample")
    print("Initializing Foundry Local Manager")
    FoundryLocalManager.initialize(config)
    manager = FoundryLocalManager.instance

    # Register execution providers
    ep_result = manager.download_and_register_eps()
    print(f"EP registration success: {ep_result.success}")

    # 2. Pick a model
    MODEL_ALIAS = "phi-4-mini"
    model = manager.catalog.get_model(MODEL_ALIAS)
    if model is None:
        print(f"Model '{MODEL_ALIAS}' not found in catalog.")
        return

    if model.is_cached:
        print(f"Model '{MODEL_ALIAS}' is already cached. Remove it first to test cancellation.")
        return

    # 3. Create a threading.Event that will signal cancellation
    cancel = threading.Event()

    # 4. Start the download on a background thread
    download_error = [None]  # mutable container for exception

    def do_download():
        try:
            model.download(
                progress_callback=lambda pct: print(f"  Download: {pct:.1f}%", end="\r"),
                cancel_event=cancel,
            )
            print("\nDownload completed successfully.")
        except Exception as e:
            download_error[0] = e

    t = threading.Thread(target=do_download)
    t.start()

    # 5. Cancel after 3 seconds
    print("Download started. Will cancel in 3 seconds...")
    time.sleep(3)
    print("\nCancelling download...")
    cancel.set()

    t.join()

    if download_error[0] is not None:
        print(f"Download was cancelled: {download_error[0]}")
    else:
        print("Download finished before cancellation took effect.")


if __name__ == "__main__":
    main()
